### PR TITLE
gh-631 - Add support for customizing OAuth2FeignReqeustInterceptor.accessTokenProvider (injecting e.g. LoadBalancerInterceptor)

### PR DESCRIPTION
**Pull request by voytech**
**1. Added support to customize `OAuth2FeignReqeustInterceptor`'s internal `AccessTokenProvider` through `OAuth2FeignReqeustInterceptorBuilder` and `OAuth2FeignRequestInterceptorConfigurer`**. 

Before `OAuth2FeignReqeustInterceptor` is created and injected into context, instances of `OAuth2FeignRequestInterceptorConfigurer` are resolved and used to customize properties of internal `AccessTokenProvider`. 

**2. Extended `FeignAutoConfiguration` so that it can create  `OAuth2FeignReqeustInterceptor` with `RetryLoadBalancerInterceptor` or `LoadBalancerInterceptor` (depends what bean is already managed).** 

IMPORTANT: Load balancing is enabled only on flag: "feign.oauth2.load-balanced=true" in order to be backward compatible, but maybe it should be default behaviour in the same way as feign load balancing.

### CODE REVIEW

File: docs/src/main/asciidoc/_configprops.adoc [hunk 1]
```diff
@@ -30,5 +30,6 @@
 |feign.httpclient.time-to-live-unit |  | 
 |feign.metrics.enabled | `true` | Enables metrics capability for Feign.
 |feign.okhttp.enabled | `false` | Enables the use of the OK HTTP Client by Feign.
-
-|===
\ No newline at end of file
+|feign.oauth2.enabled | `false` | Enables the use OAuth2FeignRequestInterceptor in order to resolve access token.
```
Comment (OlgaMaciaszek) [inline docs/src/main/asciidoc/_configprops.adoc:33]: Let's also add new properties to `additional-spring-configuration-metadata.json` for IDE support.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignAutoConfigurationTests.java [hunk 2]
```diff
@@ -81,6 +93,75 @@ void shouldInstantiateFeignCircuitBreakerTargeterWhenEnabledWithCustomCircuitBre
...
+				.withBean(BaseOAuth2ProtectedResourceDetails.class)
+				.withBean(CustomOAuth2FeignRequestInterceptorConfigurer.class).run(ctx -> {
+					assertOauth2FeignRequestInterceptorExists(ctx);
+					assertAccessTokenProviderInterceptorExists(ctx, BasicAuthenticationInterceptor.class);
+				});
+	}
+
+	private void assertOauth2FeignRequestInterceptorExists(ConfigurableApplicationContext ctx) {
+		OAuth2FeignRequestInterceptor interceptor = ctx.getBean(OAuth2FeignRequestInterceptor.class);
+		assertThat(interceptor).isNotNull();
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignAutoConfigurationTests.java:138]: Let's use `org.springframework.boot.test.context.assertj.AssertableApplicationContext` for this assertion instead: 
```
AssertableApplicationContext context = AssertableApplicationContext.get(() -> ctx);
		assertThat(context).hasSingleBean(OAuth2FeignRequestInterceptor.class);
```

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
Hello, @voytech. Thanks for providing this PR. It looks good to me, however, I have added 2 comments - please address them. Also, please add documentation for the new feature to `spring-cloud-openfeign.adoc`.

 Also adding a member of the Spring Security team for an additional review.

**Comment by OlgaMaciaszek** [comment]
@eleftherias could you please review since it relates to Spring Security integration?

**Comment by OlgaMaciaszek** [comment]
Fixes gh-631.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignAutoConfigurationTests.java (hunk 2 above)
Comment (voytech) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignAutoConfigurationTests.java:138]: Sure, did not know about this assertion. Will change soon.

File: docs/src/main/asciidoc/_configprops.adoc (hunk 1 above)
Comment (voytech) [reply docs/src/main/asciidoc/_configprops.adoc:33]: Sure, will update soon.

File: spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignAutoConfigurationTests.java (hunk 2 above)
Comment (voytech) [reply spring-cloud-openfeign-core/src/test/java/org/springframework/cloud/openfeign/FeignAutoConfigurationTests.java]: I have used AssertableApplicationContext for all assertions. It turned out that this additional reflection util for assertions was not needed anymore, so things got simplified a bit.

File: docs/src/main/asciidoc/_configprops.adoc (hunk 1 above)
Comment (voytech) [reply docs/src/main/asciidoc/_configprops.adoc:33]: I have added both property flags

File: docs/src/main/asciidoc/_configprops.adoc [hunk 3]
```diff
@@ -30,5 +30,6 @@
...
+|feign.oauth2.load-balanced | `false` | Enables `LoadBalancerInterceptor` / `RetryloadBalancerInterceptor` for `OAuth2FeignRequestInterceptor.accessTokenProvider` in order to use load balancing when resolving authentication token.
```
Comment (OlgaMaciaszek) [inline docs/src/main/asciidoc/_configprops.adoc:34]: Switch spelling to `load-balancing`.

File: docs/src/main/asciidoc/spring-cloud-openfeign.adoc [hunk 4]
```diff
@@ -760,6 +760,16 @@ feign.client.refresh-enabled=true
 ----
 TIP: DO NOT annotate the `@FeignClient` interface with the `@RefreshScope` annotation.
 
+=== OAuth2 Support
+OAuth2 support can be enabled by setting following flag:
+----
+feign.oauth2.enabled=true
+----
+When flag is set to true, and oauth2 client context with resource details are present, a bean of class `OAuth2FeignRequestInterceptor` will be created. Since now, before each request, interceptor will resolve required access token and include it as header.
```
Comment (OlgaMaciaszek) [inline docs/src/main/asciidoc/spring-cloud-openfeign.adoc:768]: Switch spelling to `load-balancing`. Switch from "should ensure that load balancer is on classpath" to  "should ensure that Spring Cloud LoadBalancer is on classpath" (it's the project official name).

File: spring-cloud-openfeign-core/src/main/resources/META-INF/additional-spring-configuration-metadata.json [hunk 5]
```diff
@@ -61,6 +61,18 @@
...
+		{
+			"name": "feign.oauth2.enabled",
+			"type": "java.lang.Boolean",
+			"description": "Enables feign interceptor for managing oauth2 access token.",
+			"defaultValue": "false"
+		},
+		{
+			"name": "feign.oauth2.load-balanced",
+			"type": "java.lang.Boolean",
+			"description": "Enables load balancing for oauth2 access token provider.",
```
Comment (OlgaMaciaszek) [inline spring-cloud-openfeign-core/src/main/resources/META-INF/additional-spring-configuration-metadata.json:74]: Change spelling to `load-balancing`.

**Review by OlgaMaciaszek — CHANGES_REQUESTED** [review]
@voytech Thanks for adding the changes. Have added some cosmetic comments - please address them. The rest looks good to me. Will also add @Buzzardo as a reviewer to take a look at the doc section.

File: docs/src/main/asciidoc/_configprops.adoc (hunk 3 above)
Comment (Buzzardo) [reply docs/src/main/asciidoc/_configprops.adoc:34]: Change the description to:

Enables `LoadBalancerInterceptor` or `RetryloadBalancerInterceptor` for `OAuth2FeignRequestInterceptor.accessTokenProvider` to use load balancing when resolving authentication token.

"load balancing" needs to be hyphenated when serving as an adjective ("load-balancing operations"). The rest of the time, it should be unhyphenated. I think we have been inconsistent about it, and I'm sure other products are inconsistent, too. But let's get it right here.

File: docs/src/main/asciidoc/spring-cloud-openfeign.adoc (hunk 4 above)
Comment (Buzzardo) [reply docs/src/main/asciidoc/spring-cloud-openfeign.adoc:768]: Change to:

When the flag is set to `true` and the oauth2 client context resource details are present, a bean of class `OAuth2FeignRequestInterceptor` is created. Before each request, the interceptor resolves the required access token and includes it as a header.

File: docs/src/main/asciidoc/spring-cloud-openfeign.adoc [hunk 6]
```diff
@@ -760,6 +760,16 @@ feign.client.refresh-enabled=true
...
+Sometimes when load balancing is enabled for Feign clients, You may want to use load balancing for fetching access token too. To achieve this, You should ensure that load balancer is on classpath (`spring-cloud-starter-loadbalancer`), as well as explicitly enable load balancing for `OAuth2FeignRequestInterceptor` by setting following flag:
```
Comment (Buzzardo) [inline docs/src/main/asciidoc/spring-cloud-openfeign.adoc:769]: Change to:

Sometimes, when load balancing is enabled for Feign clients, you may want to use load balancing for fetching access tokens, too. To do so, you should ensure that the load balancer is on the classpath (`spring-cloud-starter-loadbalancer`) and explicitly enable load balancing for `OAuth2FeignRequestInterceptor` by setting the following flag:

**Review by Buzzardo — CHANGES_REQUESTED** [review]

**Review by eleftherias — APPROVED** [review]

**Review by OlgaMaciaszek — COMMENTED** [review]
@voytech Thanks for adding the changes. This comment has not been fully addressed yet: https://github.com/spring-cloud/spring-cloud-openfeign/pull/642#discussion_r765936661
When this is done, we will be able to get this merged.

**Comment by voytech** [comment]
@OlgaMaciaszek Done.

**Pull request merged**